### PR TITLE
CORDA-3180: Added ability to lookup the associated UUID for a public key to KeyManagementService

### PR DESCRIPTION
**Pull request by roastario**

### CODE REVIEW

**Comment by roastario** [comment]
The only reason we (mid-stack) did not implement this ourselves was because there was James' work coming in the pipeline. Product are the ones asking us for the feature (token selection based on UUID) which requires this. 

I will leave it to them and the kernel team to decide whether it's worth doing.

**Comment by JamesHR3** [comment]
I appreciate that there's a desire to get this API done because it's needed for accounts, and using my implementation to implement the API makes a lot of sense. However, as @shamsasari says, we're adding to the public API by doing this, and so I think we need to make sure that the public API services any potential use cases beyond that of accounts. It would be unfortunate if we miss an opportunity to provide a better API in the core because one particular use case needs it urgently.

I've thrown together a document that proposes a slightly more fleshed out version of the API. The idea is to provide something that can still be done relatively quickly for the accounts work, but leaves open the possibility of expanding the data that can be retrieved for a key in the future. (For example, if in the future we'd like the API to tell us the whether a key is a well known identity key or an anonymous key, it should be possible to add that.)

I've PR'd this as #5414.

**Comment by roger3cev** [comment]
What's the timeline on getting this PR merged? @JamesHR3 @shamsasari - don't mean to be pushy but it's blocking accounts therefore blocking 4.3.

What @roastario has done looks alright to me. Bear in mind that we can deprecate APIs and make breaking changes in future major version bumps.

File: core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt [hunk 1]
```diff
@@ -90,4 +90,7 @@ interface KeyManagementService {
      */
     @Suspendable
     fun sign(signableData: SignableData, publicKey: PublicKey): TransactionSignature
+
+    @Suspendable
+    fun externalIdForPublicKey(pk: PublicKey): UUID?
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt]: Can we have KDocs please, in particular you need to state that a null return value can either mean the public key is unknown, or that it has no mapping.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt]: Rename `pk` to `publicKey`

**Comment by roger3cev** [comment]
Thanks @shamsasari !!

**Comment by gendal** [comment]
"Bear in mind that we can deprecate APIs and make breaking changes in future major version bumps."

This is not quite correct. It is something we have considered but it is not (yet) policy.  And, if we ever were to allow any specific compatibility break in the future, it would be because it unlocked a massive benefit that justified the impact on customers (eg to undo an insurmountable problem preventing implementation of a major new feature caused by a rushed choice in the run up to Corda 1.0, perhaps). It would not be something we could rely on to justify arguments of the form of "it's OK adding this API because we can always remove/break it at some point in the future".  The commitment to not breaking people's apps if they use committed APIs is sacrosanct, only to be breached in painful extremis.

**Pull request merged**